Elasticsearch - add 'source' argument to all functions with a 'body' argument

onlyanegg opened the pull request:
### What does this PR do?
This add support for a 'source' argument to all of the elasticsearch module functions with a 'body' argument. This way we can specify (eg.) an index template using the salt, http, or file protocols. 

### What issues does this PR fix or reference?
None

### Tests written?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

onlyanegg committed:
Adding source option to all functions with a 'body' argument (alias_create, alias_delete, document_create, index_create, index_template_create, and mapping_create)

salt-jenkins commented:
@onlyanegg, thanks for your PR! By analyzing the history of the files in this pull request, we identified @bechtoldt, @DarkKnightCZ and @msteed to be potential reviewers.

DarkKnightCZ commented:
@onlyanegg Great work, I've briefly looked at the changes and maybe two changes would be nice.

1. If somebody relies on positioning of arguments instead of names, these changes will break their code. Adding the param as last one will fix this
2. Instead of duplicating the cache logic, [cp.get_file_ str](https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.cp.html#salt.modules.cp.get_file_str) should be used (it will fetch the file, handle that cache stuff and return the content)


Please also check if the body is mandatory (otherwise not specifying body nor source should not raise exception). Nice touch would be adding those params into elasticsearch state as well.

Thanks

rallytime reviewed (changes requested):
@onlyanegg Thank you for adding this functionality. I agree with @DarkKnightCZ excellent feedback. In addition to making the changes suggested above, can you also add documentation for the new `source` options for the functions that have this new kwarg?

onlyanegg commented:
Thanks for you feedback @DarkKnightCZ and @rallytime! I definitely agree with all of those changes. 

One thing I totally forgot about is the `template` argument, which I feel goes hand in hand with the `source` argument. Should I tack that on to this PR?

Also, I was just thinking this morning after reading your comments, is there a stylistic reason this wasn't included in the beginning? I was wondering if it's more appropriate to just have the `source` and `template` arguments in the state rather than in the execution module as well.

DarkKnightCZ commented:
@onlyanegg In my opinion, including `template` would also require to include `context` as well, which is duplicating too much of `file.managed` I think (and the functionality can be achieved by just using that before Elasticsearch states). However it should be possible to implement by using something similar to [file.get_managed](https://github.com/saltstack/salt/blob/develop/salt/modules/file.py#L4012). I guess @rallytime could comment.

About second question, I guess I just forgot to include `source` when this was created, since my use-cases were different (If body is passed, then it comes from pillars and doesn't change). As the execution module can be used directly, It would be good to have `source` also available there (and not only in state).

onlyanegg committed:
Using cp.get_file_string from the cp execution module rather than the fileclient library. Adding documentation. Moving the source parameter to the end of the arguments to ensure compatibility with calls that use positional arguments.

onlyanegg commented:
@DarkKnightCZ, I'm not sure what you mean by one of your comments above.

> Please also check if the body is mandatory (otherwise not specifying body nor source should not raise exception).

I think, as it is, an exception will not be raised if neither body nor source are specified which I believe is the correct behavior.

DarkKnightCZ commented:
@onlyanegg Ah, my bad, misread that condition.

onlyanegg commented:
Ok, I moved the source argument to the end, am now using the `cp.get_file_str` function (much easier :), and added documentation. I'll leave out the template argument for now, and maybe look into `file.get_managed` later, and I'll also forgo the source argument in the state for now. 

I also completely fucked up my branch and ended up force pushing 150 odd commits, but then I created a new local branch, cherry picked my commits, and force pushed it. So I think everything is ok now.

cachedout commented on salt/modules/elasticsearch.py:
```diff
@@ -373,7 +389,13 @@ def document_create(index, doc_type, body=None, id=None, hosts=None, profile=Non
     es = _get_instance(hosts, profile)
-
+    if source and body:
+        message = 'Either body or source should be specified but not both.'
```
IMHO, we need to document these sorts of limitation in the function documentation.

onlyanegg replied:
ok, that's a good idea

cachedout commented on salt/modules/elasticsearch.py:
```diff
@@ -455,22 +477,30 @@ def document_get(index, id, doc_type='_all', hosts=None, profile=None):
     es = _get_instance(hosts, profile)
-
+    if source and body:
+        message = 'Either body or source should be specified but not both.'
```
Same here. Need docs.

onlyanegg committed:
Adding more documentation

cachedout commented:
@rallytime Re-review requsted.

rallytime reviewed (approved):
Much cleaner. Very good!
